libvast/include/vast/detail/series_builders.hpp

```diff
@@ -291,6 +292,27 @@ struct series_builder : series_builder_base {
                    new_builder.add(view);
                    *this = std::move(new_builder);
                  },
+                 [view]<class BuilderType>(
+                   concrete_series_builder<BuilderType>& builder) {
+                   using BuilderView = type_to_data_t<BuilderType>;
+                   if constexpr (std::is_same_v<Type, int64_type>
+                                 and std::is_same_v<BuilderType, uint64_type>) {
+                     if (view < 0u) {
+                       die("trying to add a negative value into uint64 type "
+                           "builder");
+                     }
+                     builder.add(static_cast<BuilderView>(view));
+                   }
+                   if constexpr (std::is_same_v<Type, uint64_type>
+                                 and std::is_same_v<BuilderType, int64_type>) {
+                     if (view > std::numeric_limits<BuilderView>::max()) {
+                       die(
+                         "trying to add an uint64_t value that is higher than "
+                         "int_64_t max");
+                     }
+                     builder.add(static_cast<BuilderView>(view));
+                   }
+                 },
```

Comment:
Is this perhaps something we can express more generally? Should the `series_builder<T>::add` API not do this internally?